File: src/main/java/xyz/phanta/cbtweaker/hatch/HatchType.java
package xyz.phanta.cbtweaker.hatch;

import net.minecraft.block.state.IBlockState;
import net.minecraft.item.ItemStack;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import net.minecraftforge.fml.common.registry.ForgeRegistries;
import xyz.phanta.cbtweaker.CbtMod;
import xyz.phanta.cbtweaker.buffer.BufferFactory;
import xyz.phanta.cbtweaker.buffer.BufferObserver;
import xyz.phanta.cbtweaker.buffer.BufferType;
import xyz.phanta.cbtweaker.common.BlockMaterial;
import xyz.phanta.cbtweaker.gui.inventory.UiLayout;

import javax.annotation.Nullable;
import java.util.List;
import java.util.Objects;

public class HatchType<B> {

    private final String id;
    private final BlockMaterial blockMaterial;
    private final BufferType<B, ?, ?, ?> bufferType;
    private final List<TierData<B>> tierData;

    @Nullable
    private HatchBlock hatchBlock = null;

    public HatchType(String id, BlockMaterial blockMaterial,
                     BufferType<B, ?, ?, ?> bufferType, List<TierData<B>> tierData) {
        this.id = id;
        this.blockMaterial = blockMaterial;
        this.bufferType = bufferType;
        this.tierData = tierData;
    }

    public String getId() {
        return id;
    }

    public BlockMaterial getBlockMaterial() {
        return blockMaterial;
    }

    public BufferType<B, ?, ?, ?> getBufferType() {
        return bufferType;
    }

    public int getTierCount() {
        return tierData.size();
    }

    public TierData<B> getTier(int tier) {
        return tierData.get(tier);
    }

    public HatchBlock getHatchBlock() {
        if (hatchBlock == null) {
            hatchBlock = Objects.requireNonNull(
                    (HatchBlock)ForgeRegistries.BLOCKS.getValue(CbtMod.INSTANCE.newResourceLocation("hatch_" + id)));
        }
        return hatchBlock;
    }

    public IBlockState getHatchBlock(int tier) {
        HatchBlock block = getHatchBlock();
        return block.getDefaultState().withProperty(block.getTierProperty(), tier);
    }

    public ItemStack getHatchStack(int count, int tier) {
        HatchBlock block = getHatchBlock();
        return new ItemStack(block, count, tier);
    }

    public static class TierData<B> {

        private final BufferFactory<B, ?> bufferFactory;
        private final UiLayout uiLayout;

        public TierData(BufferFactory<B, ?> bufferFactory, UiLayout uiLayout) {
            this.bufferFactory = bufferFactory;
            this.uiLayout = uiLayout;
        }

        public B createBuffer(World world, BlockPos pos, @Nullable BufferObserver observer) {
            return bufferFactory.createBuffer(world, pos, observer);
        }

        public UiLayout getUiLayout() {
            return uiLayout;
        }

    }

}

File: src/main/kotlin/st/evening/mc/cbtweaker/hatch/HatchType.kt
package st.evening.mc.cbtweaker.hatch

import net.minecraft.block.state.IBlockState
import net.minecraft.item.ItemStack
import st.evening.mc.cbtweaker.buffer.BufferFactory
import st.evening.mc.cbtweaker.buffer.BufferType
import st.evening.mc.cbtweaker.common.BlockConfig
import st.evening.mc.cbtweaker.common.CustomBlockType
import st.evening.mc.cbtweaker.gui.inventory.WindowConfig
import st.evening.mc.prelude.api.registration.ModRegistrar
import st.evening.mc.prelude.api.registration.block

class HatchType<B>(
    reg: ModRegistrar,
    val id: String,
    override val blockConfig: BlockConfig,
    val bufferType: BufferType<B, *, *, *>,
    private val tierData: List<TierData<B>>
) : CustomBlockType {
    val hatchBlock: HatchBlock by reg.block("hatch_$id") { HatchBlock.construct(this) }

    val tierCount: Int
        get() = tierData.size

    fun getTier(tier: Int): TierData<B> = tierData[tier]

    fun getHatchBlock(tier: Int): IBlockState {
        val block = hatchBlock
        return block.defaultState.withProperty(block.tierProperty, tier)
    }

    fun getHatchStack(count: Int, tier: Int): ItemStack = ItemStack(hatchBlock, count, tier)

    class TierData<B>(val bufferFactory: BufferFactory<B, *>, val windowConfig: WindowConfig)
}
